src/Client/HedgedConnectionsFactory.cpp

```diff
@@ -0,0 +1,360 @@
+
+    if (!connection_establisher.getFailMessage().empty())
+        fail_messages += connection_establisher.getFailMessage() + "\n";
+
+    ShuffledPool & shuffled_pool = shuffled_pools[index];
+    LOG_WARNING(
+        log, "Connection failed at try №{}, reason: {}", (shuffled_pool.error_count + 1), connection_establisher.getFailMessage());
+    ProfileEvents::increment(ProfileEvents::DistributedConnectionFailTry);
+
+    shuffled_pool.error_count = std::min(pool->getMaxErrorCup(), shuffled_pool.error_count + 1);
+
+    if (shuffled_pool.error_count >= max_tries)
+    {
+        ++failed_pools_count;
+        ProfileEvents::increment(ProfileEvents::DistributedConnectionFailAtAll);
+    }
+
+    --replicas_in_process_count;
+}
+
+HedgedConnectionsFactory::State HedgedConnectionsFactory::processEpollEvents(bool blocking, Connection *& connection_out)
+{
+    int event_fd;
+    while (!epoll.empty())
+    {
+        event_fd = getReadyFileDescriptor(blocking);
+
+        if (event_fd == -1)
+            return State::NOT_READY;
+
+        if (fd_to_replica_index.contains(event_fd))
+        {
+            int index = fd_to_replica_index[event_fd];
+            processConnectionEstablisherEvent(index, connection_out);
+
+            if (replicas[index].is_ready)
+                return State::READY;
+            if (replicas[index].connection_establisher.isInProcess())
+                continue;
+        }
+        else if (timeout_fd_to_replica_index.contains(event_fd))
+            replicas[timeout_fd_to_replica_index[event_fd]].change_replica_timeout.reset();
+        else
+            throw Exception("Unknown event from epoll", ErrorCodes::LOGICAL_ERROR);
+
+        /// We reach this point only if we need to start new connection
+        /// (Special timeout expired or one of the previous connections failed).
+        int index = startEstablishingNewConnection(connection_out);
+
+        /// Return only if replica is ready.
+        if (index != -1 && replicas[index].is_ready)
+            return State::READY;
+    }
+
+    return State::CANNOT_CHOOSE;
+}
+
+int HedgedConnectionsFactory::getReadyFileDescriptor(bool blocking)
+{
+    epoll_event event;
+    event.data.fd = -1;
+    epoll.getManyReady(1, &event, blocking);
+    return event.data.fd;
+}
+
+void HedgedConnectionsFactory::removeReplicaFromEpoll(int index)
+{
+    ReplicaStatus & replica = replicas[index];
+    epoll.remove(replica.connection_establisher.getFileDescriptor());
+    fd_to_replica_index.erase(replica.connection_establisher.getFileDescriptor());
+
+    replica.change_replica_timeout.reset();
+    epoll.remove(replica.change_replica_timeout.getDescriptor());
+    timeout_fd_to_replica_index.erase(replica.change_replica_timeout.getDescriptor());
+}
+
+void HedgedConnectionsFactory::processConnectionEstablisherEvent(int index, Connection *& connection_out)
+{
+    replicas[index].connection_establisher.resume();
+    processConnectionEstablisherStage(index, true);
```

Comment:
I think it will be much easier if remove replica from `epoll` here manually if `replica.connection_establisher.isFinished() || replica.connection_establisher.isFailed()`. Flag `remove_from_epoll` is very confusing, it is good if it may be removed.